src/main/java/org/elasticsearch/indices/store/IndicesStore.java

```diff
@@ -333,8 +336,69 @@ public String executor() {
         }
 
         @Override
-        public void messageReceived(ShardActiveRequest request, TransportChannel channel) throws Exception {
-            channel.sendResponse(new ShardActiveResponse(shardActive(request), clusterService.localNode()));
+        public void messageReceived(final ShardActiveRequest request, final TransportChannel channel) throws Exception {
+            // make sure shard is really there before register cluster state obsever
+            ShardId shardId = request.shardId;
+            IndexService indexService = indicesService.indexService(shardId.index().getName());
+            if (indexService != null && indexService.indexUUID().equals(request.indexUUID)) {
+                if (!indexService.hasShard(shardId.id())) {
+                    channel.sendResponse(new ShardActiveResponse(false, clusterService.localNode()));
+                }
+            } else {
+                channel.sendResponse(new ShardActiveResponse(false, clusterService.localNode()));
+            }
+            // create observer just in case
+            ClusterStateObserver observer = new ClusterStateObserver(clusterService, request.timeout, logger);
```

Comment:
seems to me like there is no need to close and the close() method is unused anyway. I will remove it to avoid further confusion.